### PR TITLE
Sample: Add the custom cell

### PR DESCRIPTION
**Pull request by Filastian**
The custom cell element has been added, which consists of:
* Header label;
* Value label - a random number;
* A button that generates a random number and sets it to a value.

image

### CODE REVIEW

File: Buform.Example.Core/ComponentsViewModel.cs [hunk 1]
```diff
@@ -298,6 +299,13 @@ public ComponentsViewModel(ILoggerFactory logFactory, IMvxNavigationService navi
                         Label = "Time",
                         InputType = DateTimeInputType.Time
                     }
+                },
+                new TextFormGroup("Digit generation")
+                {
+                    new DigitGenerationItem(() => GeneratedDigit)
```
Comment (prin53) [inline Buform.Example.Core/ComponentsViewModel.cs:305]: I would call this item RandomNumberGenerator or something like that, as Generation not so much info about its functionality.

File: Buform.Example.Core/ComponentsViewModel.cs [hunk 2]
```diff
@@ -298,6 +299,13 @@ public ComponentsViewModel(ILoggerFactory logFactory, IMvxNavigationService navi
...
+                    {
+                        Label = "Generated digit:"
```
Comment (prin53) [inline Buform.Example.Core/ComponentsViewModel.cs:307]: Just "Number" should be enough.

File: Buform.Example.Core/DigitGenerationItem.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public sealed class DigitGenerationItem : FormItem<int>
+    {
+        private const int MinValue = 0;
+        private const int MaxValue = 1000;
+
+        private readonly Random _random;
+
+        public string? Label { get; set; }
+        public ICommand RegenerateCommand { get; }
```
Comment (prin53) [inline Buform.Example.Core/DigitGenerationItem.cs:16]: Are we regenerating something here? I think it should be just `GenerateCommand`

File: Buform.Example.Core/ComponentsViewModel.cs [hunk 4]
```diff
@@ -298,6 +299,13 @@ public ComponentsViewModel(ILoggerFactory logFactory, IMvxNavigationService navi
                         Label = "Time",
                         InputType = DateTimeInputType.Time
                     }
+                },
+                new TextFormGroup("Digit generation")
```
Comment (prin53) [inline Buform.Example.Core/ComponentsViewModel.cs:303]: I would move it to the menu form, as it demonstrates the usage of a custom header. And the components form is for embedded controls only.

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        private void ExecuteCommand(object _, EventArgs __)
+        {
+            Item?.RegenerateCommand.SafeExecute();
+        }
+
+        protected override void Initialize()
+        {
+            SelectionStyle = UITableViewCellSelectionStyle.None;
+            ContentView.UserInteractionEnabled = true;
```
Comment (prin53) [inline Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs:73]: I don't think we need this line here.

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            _valueLabel = new UILabel
+            {
+                TranslatesAutoresizingMaskIntoConstraints = false,
+                Font = UIFont.PreferredBody,
+                TextColor = UIColor.SecondaryLabel
+            };
+
+            _button = UIButton.FromType(UIButtonType.System);
+            _button.Frame = new CGRect(0, 0, 70, 40);
+            _button.TouchUpInside += ExecuteCommand;
```
Comment (prin53) [inline Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs:91]: The handler method must be called like the handler itself, but with `On` prefix to clearly indicate the connected event.

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+            _valueLabel = new UILabel
+            {
+                TranslatesAutoresizingMaskIntoConstraints = false,
+                Font = UIFont.PreferredBody,
+                TextColor = UIColor.SecondaryLabel
+            };
+
+            _button = UIButton.FromType(UIButtonType.System);
+            _button.Frame = new CGRect(0, 0, 70, 40);
```
Comment (prin53) [inline Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs:90]: Not sure if we need to hardcode frame here?

**Review by prin53 — CHANGES_REQUESTED** [review]
I would also follow the main pattern in the UI implementation:
- The title label should be aligned to the left part
- The value label should be aligned to the left part
- The accessory comes after the value label

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs (hunk 5 above)
Comment (Filastian) [reply Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs:73]: I have added this one during development and forgot to remove it, thanks 👍

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs (hunk 7 above)
Comment (Filastian) [reply Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs:90]: But if we don't specify the frame property, the button will be non displayed.

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs (hunk 6 above)
Comment (prin53) [reply Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs]: `button.TouchUpInside += OnButtonTouched` -> `button.TouchUpInside += OnTouchUpInside`

File: Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs (hunk 7 above)
Comment (prin53) [reply Buform.Example.MvvmCross.iOS/DigitGenerationCell.cs]: I would use the default value, like 36x36 then.

File: Buform.Example.Core/DigitGenerationItem.cs (hunk 3 above)
Comment (prin53) [reply Buform.Example.Core/DigitGenerationItem.cs]: The command callback function should be named by the command name as well to be clear. E.g. Regenerate -> Generate

**Review by prin53 — CHANGES_REQUESTED** [review]
See my comments above.

Some new things:
- Sorry, I wrote the UI changes wrongly: the number label should be aligned to the right, before the accessory button
- Please ensure that class names correspond with the file names

**Comment by Filastian** [comment]
I have replaced the hardcoded size with the image size and now it looks like that:
image

**Review by prin53 — APPROVED** [review]

**Pull request merged**